tests/src/core/testqgsprojectstorage.cpp

```diff
@@ -0,0 +1,221 @@
+
+      QStringList lst = uri.split( ":" );
+      Q_ASSERT( lst.count() == 2 );
+      QString projectName = lst[1];
+
+      if ( !mProjects.contains( projectName ) )
+      {
+        context.pushMessage( "project not found", Qgis::Critical );
+        return false;
+      }
+
+      QByteArray content = mProjects[projectName];
+      ioDevice->write( content );
+      ioDevice->seek( 0 );
+      return true;
+    }
+
+    virtual bool writeProject( const QString &uri, QIODevice *ioDevice, QgsReadWriteContext &context ) override
+    {
+      Q_UNUSED( context );
+      QStringList lst = uri.split( ":" );
+      Q_ASSERT( lst.count() == 2 );
+      QString projectName = lst[1];
+
+      mProjects[projectName] = ioDevice->readAll();
+
+      QgsProjectStorage::Metadata meta;
+      meta.name = projectName;
+      meta.lastModified = QDateTime::currentDateTime();
+      mProjectsMetadata[projectName] = meta;
+      return true;
+    }
+
+    virtual bool removeProject( const QString &uri ) override
+    {
+      QStringList lst = uri.split( ":" );
+      Q_ASSERT( lst.count() == 2 );
+      QString projectName = lst[1];
+
+      if ( !mProjects.contains( projectName ) )
+        return false;
+
+      mProjects.remove( projectName );
+      mProjectsMetadata.remove( projectName );
+      return true;
+    }
+
+    virtual bool readProjectMetadata( const QString &uri, QgsProjectStorage::Metadata &metadata ) override
+    {
+      QStringList lst = uri.split( ":" );
+      Q_ASSERT( lst.count() == 2 );
+      QString projectName = lst[1];
+      if ( !mProjects.contains( projectName ) )
+        return false;
+
+      metadata = mProjectsMetadata[projectName];
+      return true;
+    }
+
+  private:
+    QHash<QString, QByteArray> mProjects;
+    QHash<QString, QgsProjectStorage::Metadata> mProjectsMetadata;
+};
+
+
+void TestQgsProjectStorage::testMemoryStorage()
+{
+  QString dataDir( TEST_DATA_DIR ); // defined in CmakeLists.txt
+  QString layerPath = dataDir + "/points.shp";
+  QgsVectorLayer *layer1 = new QgsVectorLayer( layerPath, QStringLiteral( "points" ), QStringLiteral( "ogr" ) );
+  QVERIFY( layer1->isValid() );
+
+  MemoryStorage *memStorage = new MemoryStorage;
+
+  QgsApplication::projectStorageRegistry()->registerProjectStorage( memStorage );
+
+  QVERIFY( QgsApplication::projectStorageRegistry()->projectStorages().contains( memStorage ) );
+
+  QCOMPARE( memStorage->listProjects( QString() ).count(), 0 );
+
+  QgsProject prj1;
+  prj1.setTitle( "best project ever" );
+  prj1.addMapLayer( layer1 );
+  prj1.setFileName( "memory:project1" );
+  bool writeOk = prj1.write();
+
+  QVERIFY( writeOk );
+  QCOMPARE( memStorage->listProjects( QString() ).count(), 1 );
+
+  QVERIFY( prj1.absoluteFilePath().isEmpty() );
+  QCOMPARE( prj1.baseName(), QString( "project1" ) );
+  QVERIFY( prj1.lastModified().secsTo( QDateTime::currentDateTime() ) < 1 );
+
+  // read the project back
```

Comment:
Can you add a test here which uses zipped contents too?

Reply:
Good point - added in b8a9329